### ISSUE TITLE
[no-invalid-void-type] Intersection with `never` should be allowed

### ISSUE THREAD

**Issue by jens-duttke** [typescript-eslint/typescript-eslint#1946]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": "error"
  }
}
```

```TS
function foo (): void | never {
    throw new Error('Test');
}
```

**Expected Result**

That the **no-invalid-void-type** does not complain about the `void | never` union.

**Actual Result**

> void is only valid as a return type or generic type variable [Error/@typescript-eslint/no-invalid-void-type]

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.30.0` |
| `@typescript-eslint/parser`        | `2.30.0` |
| `TypeScript`                       | `3.9.0-dev.20200419` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.6.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
How come you're using `void | never`?
TypeScript collapses the `never` out of the return type, so it's like it's never even there.

Is it purely for documentation that the function may throw?

**Comment by jens-duttke** [comment]
@bradzacher: Yes, right now it's purely for documentation. But I also have the hope that some day, either TypeScript or an ESLint plugin does something useful with it. ;-)

**Comment by sinyovercosy** [comment]
This should be an easy fix once #1960 is merged - all that should be needed is to add `never` as a valid member type to the `isValidUnionType` function.

**Comment by auhsoja** [comment]
There are no other union-type signatures for which this exception should be made, right? Seems like it'd be bad style and/or unsafe in most other cases, but I haven't thought about it for very long.
